www/app/app.ts

```diff
@@ -28,6 +28,32 @@ import {Settings, SettingsKey} from './settings';
 import {Updater} from './updater';
 import {UrlInterceptor} from './url_interceptor';
 
+// If s a URL whose fragment contains a Shadowsocks URL then return that Shadowsocks URL,
+// otherwise return s.
+function unwrapInvite(s: string): string {
+  try {
+    const url = new URL(s);
+    if (url.hash) {
+      const decodedFragment = decodeURIComponent(url.hash);
+
+      // Search in the fragment for ss:// for two reasons:
+      //  - URL.hash includes the leading # (what).
+      //  - When a user opens invite.html#ENCODEDSSURL in their browser, the website (currently)
+      //    redirects to invite.html#/en/invite/ENCODEDSSURL. Since that seems like a reasonable
+      //    thing to do, let's support those URLs too.
+      const possibleShadowsocksUrl = decodedFragment.substring(decodedFragment.indexOf('ss://'));
+
+      if (new URL(possibleShadowsocksUrl).protocol === 'ss:') {
+        return possibleShadowsocksUrl;
```

Comment:
It seems like the net effect here is to make `unwrapInvite` much less strict, and let `confirmAddServer` catch any errors.  That seems likely to work, and should achieve the goal of reducing log spam as long as `confirmAddServer` and `SHADOWSOCKS_URI.parse` don't log on failure when reading from the clipboard.

The only case I can think of where this will truly fail for us is if a server operator, using the fragment as a free text field, includes `ss://` in the server description, e.g. `ss://...#Foonet, your source for crummy ss:// links`.  If you're concerned about this case, you could add a check for `url.protocol === 'ss:'` and bail out early.

I do wonder how this change will affect our ability to evolve the invite format going forward.  For example, if a future invite is no longer compatible with old client versions (e.g. no UDP support), we could currently indicate that by using a different URL or fragment prefix.  This change would remove that capability.  Conversely, this change would allow us to create backward-compatible invites that also contain a new (e.g. non-shadowsocks) access key type, provided that the shadowsocks access key is at the end of the fragment.

Reply:
Good points.

I suspect the ship has sailed on backwards compatibility, given we already have clients out there that will greedily consume any and all ss:// links. Usually there's a way to hack together some kind of backwards compatibility if you really have to (also, the app has to explicitly register the URL paths it's interested in - I see manually pasting a URL with a fragment very much as an edge case...).